lacommunaute/users/management/commands/populate_emaillastseen.py

```diff
@@ -0,0 +1,125 @@
+        .values_list("voter__email", "created_at", "kind")
+    )
+    return list(qs)
+
+
+def collect_forum_rating():
+    qs = (
+        ForumRating.objects.exclude(user=None)
+        .annotate(kind=Value(EmailLastSeenKind.FORUM_RATING))
+        .values_list("user__email", "created", "kind")
+    )
+    return list(qs)
+
+
+def collect_post():
+    qs_authenticated = (
+        Post.objects.exclude(poster=None)
+        .annotate(kind=Value(EmailLastSeenKind.POST))
+        .values_list("poster__email", "created", "kind")
+    )
+    qs_anonymous = (
+        Post.objects.filter(poster=None)
+        .annotate(kind=Value(EmailLastSeenKind.POST))
+        .values_list("username", "created", "kind")
+    )
+    return list(qs_authenticated) + list(qs_anonymous)
+
+
+def collect_clicked_notifs():
+    # TODO VincentPorte, en attente #891
+    sys.stdout.write("collect_clicked_notifs: pending #891\n")
+    return []
+
+
+def deduplicate(last_seen):
+    return {tup[0]: tup for tup in sorted(last_seen, key=lambda tup: (tup[0], tup[1]))}
+
+
+def remove_known_last_seen(dedup_last_seen_dict):
+    known_last_seen = EmailLastSeen.objects.values_list("email", flat=True)
+    return {k: v for k, v in dedup_last_seen_dict.items() if k not in known_last_seen}
+
+
+def insert_last_seen(dedup_last_seen_dict):
+    obj = [EmailLastSeen(email=k, last_seen_at=v[1], last_seen_kind=v[2]) for k, v in dedup_last_seen_dict.items()]
+    return EmailLastSeen.objects.bulk_create(obj, batch_size=1000)
+
+
+class Command(BaseCommand):
+    help = "hydratation de la table EmailLastSeen avec la date de dernière visite des emails connus"
+
+    def handle(self, *args, **options):
+        last_seen = collect_users_logged_in()
+        sys.stdout.write(f"users logged in: collected {len(last_seen)}\n")
+
+        last_seen += collect_event()
+        sys.stdout.write(f"events: collected {len(last_seen)}\n")
+
+        last_seen += collect_DSP()
+        sys.stdout.write(f"DSP: collected {len(last_seen)}\n")
+
+        last_seen += collect_upvote()
+        sys.stdout.write(f"UpVotes: collected {len(last_seen)}\n")
+
+        last_seen += collect_forum_rating()
+        sys.stdout.write(f"forum ratings: collected {len(last_seen)}\n")
+
+        last_seen += collect_post()
+        sys.stdout.write(f"posts: collected {len(last_seen)}\n")
+
+        last_seen += collect_clicked_notifs()
+        sys.stdout.write(f"clicked notifications: collected {len(last_seen)}\n")
+
+        dedup_last_seen_dict = deduplicate(last_seen)
+        sys.stdout.write(f"deduplication: {len(dedup_last_seen_dict)}\n")
+
+        dedup_last_seen_dict = remove_known_last_seen(dedup_last_seen_dict)
+        sys.stdout.write(f"remove known last seen: {len(dedup_last_seen_dict)}\n")
+
+        res = insert_last_seen(dedup_last_seen_dict)
+        sys.stdout.write(f"insert last seen: {len(res)}\n")
+
+        sys.stdout.write("that's all folks!\n")
+        sys.stdout.flush()
```

Comment:
C’est violent niveau mémoire de tout charger, mais vu la taille de la commu :shrug: 

J’aurais plutôt utilisé l’identifiant utilisateur comme clé d’un `dict[username, namedtuple(last_seen, kind)]` et itéré sur les éléments petit à petit.

Reply:
Il y a aussi l'approche que j'ai utilisé pour la création des groupes de suivi GPS -> j'annote les users et je fais des petits batchs aussi : 
https://github.com/gip-inclusion/les-emplois/blob/master/itou/gps/management/commands/sync_follow_up_groups_and_members.py
(je dis plus ça pour la culture que pour recommander de faire pareil)